### ISSUE TITLE
Adding meta for routes without a component

### ISSUE THREAD

**Issue by posva** [posva/unplugin-vue-router#33]
Creating folders create routes and any file inside them becomes a nested route. This wasn't possible before v4.1 but now it makes sense to allow defining meta properties (mainly but also any other route property) at this level

This could maybe be achieved with a special file `users/__route.js` that exports some properties like name, meta, etc (the same as in the `<route>` block)

**Comment by Soviut** [comment]
This would be nice for applying `meta: { requiresAuth: true }` to a parent route. Currently, I'm having to use `definePage` on every child page.

**Comment by jods4** [comment]
Here we're building our navigation menus from the routes table (using metadata added in route blocks).
We'd like to add metadata to the non-navigable parent nodes, too.

We used the following hack to work-around this:
1. Create an empty `index.vue` page anyway, with a `<route>` block and metadata.
2. In the metadata, include a marker that the page must be removed, e.g. `"remove": true`.
3. We use the `extendRoute(route)` callback during build to customize the process a bit:
```js
if (route.node.value.overrides?.remove)
{
  // Move metadata to parent, depends on your use cases
  route.parent.addToMeta(route.node.value.overrides.meta);
  // Remove from routing table
  route.delete();
}
```

If anyone's interested, that approach works.
It would be nice to have this built-in though, maybe as the default export of `__route.js` as suggested initially.

Random remark: `route.node` is described as a private variable in code, yet there is no access to `overrides` object from `EditableTreeNode`. I feel like one public API is missing?

**Comment by posva** [comment]
Thanks for sharing your approach!
I'm realizing this is quite old now and I haven't implemented any experimental APIs around it yet... I hope to get back to it after #246 and the Data Loaders RFC

> Random remark: route.node is described as a private variable in code, yet there is no access to overrides object from EditableTreeNode. I feel like one public API is missing?

Yes, EditableTreeNode is definitely still missing some APIs but also docs and examples 😅

**Comment by an600027** [comment]
> Here we're building our navigation menus from the routes table (using metadata added in route blocks). We'd like to add metadata to the non-navigable parent nodes, too.
> 
> We used the following hack to work-around this:
> 
> 1. Create an empty `index.vue` page anyway, with a `<route>` block and metadata.
> 2. In the metadata, include a marker that the page must be removed, e.g. `"remove": true`.
> 3. We use the `extendRoute(route)` callback during build to customize the process a bit:
> 
> ```js
> if (route.node.value.overrides?.remove)
> {
>   // Move metadata to parent, depends on your use cases
>   route.parent.addToMeta(route.node.value.overrides.meta);
>   // Remove from routing table
>   route.delete();
> }
> ```
> 
> If anyone's interested, that approach works. It would be nice to have this built-in though, maybe as the default export of `__route.js` as suggested initially; **or `__route.json` and `__route.yaml` for syntax consistency with `<route>` blocks.**
> 
> Random remark: `route.node` is described as a private variable in code, yet there is no access to `overrides` object from `EditableTreeNode`. I feel like one public API is missing?

`route["node"].value.overrides.remove`
这样调用后编译时期不会报错，When called like this, no errors are reported at compile time

**Comment by an600027** [comment]
> Here we're building our navigation menus from the routes table (using metadata added in route blocks). We'd like to add metadata to the non-navigable parent nodes, too.
> 
> We used the following hack to work-around this:
> 
> 1. Create an empty `index.vue` page anyway, with a `<route>` block and metadata.
> 2. In the metadata, include a marker that the page must be removed, e.g. `"remove": true`.
> 3. We use the `extendRoute(route)` callback during build to customize the process a bit:
> 
> ```js
> if (route.node.value.overrides?.remove)
> {
>   // Move metadata to parent, depends on your use cases
>   route.parent.addToMeta(route.node.value.overrides.meta);
>   // Remove from routing table
>   route.delete();
> }
> ```
> 
> If anyone's interested, that approach works. It would be nice to have this built-in though, maybe as the default export of `__route.js` as suggested initially; **or `__route.json` and `__route.yaml` for syntax consistency with `<route>` blocks.**
> 
> Random remark: `route.node` is described as a private variable in code, yet there is no access to `overrides` object from `EditableTreeNode`. I feel like one public API is missing?

**Called like this, vscode does not report an error**
`route["node"].value.overrides.remove`

**Comment by AAABingBing** [comment]
> Creating folders create routes and any file inside them becomes a nested route. This wasn't possible before v4.1 but now it makes sense to allow defining meta properties (mainly but also any other route property) at this level
> 
> This could maybe be achieved with a special file `users/__route.js` that exports some properties like name, meta, etc (the same as in the `<route>` block)

How to write the `__route.js` file

**Comment by jods4** [comment]
@AAABingBing This issue is open, there is no support for `__route.js` (or any other alternative) built into this library yet.
Look at my comment above for an alternative using `extendRoute()` to implement this feature in your code.

**Comment by AAABingBing** [comment]
> @AAABingBing This issue is open, there is no support for `__route.js` (or any other alternative) built into this library yet. Look at my comment above for an alternative using `extendRoute()` to implement this feature in your code.

Oh, I see
Thank you for your reply

**Comment by AAABingBing** [comment]
> Here we're building our navigation menus from the routes table (using metadata added in route blocks). We'd like to add metadata to the non-navigable parent nodes, too.
> 
> We used the following hack to work-around this:
> 
> 1. Create an empty `index.vue` page anyway, with a `<route>` block and metadata.
> 2. In the metadata, include a marker that the page must be removed, e.g. `"remove": true`.
> 3. We use the `extendRoute(route)` callback during build to customize the process a bit:
> 
> ```js
> if (route.node.value.overrides?.remove)
> {
>   // Move metadata to parent, depends on your use cases
>   route.parent.addToMeta(route.node.value.overrides.meta);
>   // Remove from routing table
>   route.delete();
> }
> ```
> 
> If anyone's interested, that approach works. It would be nice to have this built-in though, maybe as the default export of `__route.js` as suggested initially; **or `__route.json` and `__route.yaml` for syntax consistency with `<route>` blocks.**
> 
> Random remark: `route.node` is described as a private variable in code, yet there is no access to `overrides` object from `EditableTreeNode`. I feel like one public API is missing?

This method solves my problem, but I have another question
Using `definePage` configuration, unable to use `route.node.value.overrides` to access route information
But in some cases I have to use the `definePage` configuration route information

Such as:

![image](https://github.com/user-attachments/assets/d5a48cb9-e806-4028-a51a-c4f0f73cc1b5)

In this case, I cannot use the `route` block to configure

![image](https://github.com/user-attachments/assets/e5d984f4-d4a8-4c2c-b4ca-9f4a0ccb017a)

**Comment by posva** [comment]
@AAABingBing in your case, use `definePage()` and add a pass through template:

```vue
<RouterView />
```

That will add `component: () => h(RouterView)` which is like not having a component.

**Comment by AAABingBing** [comment]
> @AAABingBing in your case, use `definePage()` and add a pass through template:
> 
> ```
> <RouterView />
> ```
> 
> That will add `component: () => h(RouterView)` which is like not having a component.

Sorry, maybe I understand incorrectly

```vue
<script setup>
import IconLayout from '~icons/tabler/layout'

definePage({
  remove: true,
  meta: {
    title: 'Test',
    icon: IconLayout,
  },
})
</script>

<template>
  <RouterView />
</template>
```

This way still unable to use `route.node.value.overrides` to access route information
It is still necessary to use the `route` block

```js
VueRouter({
  extendRoute(route) {
    // if (route.node.value.overrides?.remove) {
    //   route.parent.addToMeta(route.node.value.overrides.meta)
    //   route.delete()
    // }
  },
  dts: 'src/typed-router.d.ts',
}),
```

**Comment by posva** [comment]
Just open a discussion with all the information. I personally cont understand what you are trying to do

**Comment by AAABingBing** [comment]
I use the above way to add metadata to the non-navigable parent nodes
But I found that this method must use `route` blocks instead of `definePage()` instead

Just like this, i don’t know how to use  imported `IconLayout` icon in the `route` block:

![image](https://github.com/user-attachments/assets/39a1a57f-1c3f-449a-9f24-0de79199eb67)

**Comment by AAABingBing** [comment]
> Just open a discussion with all the information. I personally don't understand what you are trying to do

I see
Thank you for your reply

**Comment by rickiewars** [comment]
For anyone interested, I’d like to share my temporary workaround for this issue. In this example, you can add a `route.js` file to any folder where you want to recursively apply certain metadata or settings.

Add the following to your `vite.config.ts` file under `plugins.VueRouter`:
```php
async beforeWriteFiles(rootRoute) {
  const forEachRoute = (
    route: EditableTreeNode,
    callback: (route: EditableTreeNode) => Promise<void>
  ) => {
    callback(route)
    if (route.children) {
      route.children.forEach((childRoute: EditableTreeNode) => {
        forEachRoute(childRoute, callback)
      })
    }
  }

  type ExtendRouteData = {
    route: EditableTreeNode
    data: any
  }
  const promises: Array<Promise<ExtendRouteData | null>> = []

  const loadExtendedRoute = async (route: EditableTreeNode) => {
    const pagesDir = fileURLToPath(new URL('./resources/pages', import.meta.url))
    const routeFile = path.join(pagesDir, route.fullPath, 'route.js')
    promises.push(
      import(routeFile)
        .then((module) => {
          return { route: route, data: module.default }
        })
        .catch(() => null)
    )
  }

  forEachRoute(rootRoute, async (route: EditableTreeNode) => {
    if (route.isPassThrough) {
      loadExtendedRoute(route)
    }
  })

  await Promise.all(promises).then((results) => {
    results.forEach((result) => {
      if (result) {
        result.route.addToMeta(result.data)
      }
    })
  })
}
```

This approach is not without its drawbacks. Since unplugin-router doesn't monitors the `route.js` files to regenerate routes, changes made to the `route.js` files won’t be detected unless Vite is restarted. Additionally, it would be ideal to define these route overrides in a TypeScript file.

For now, this is a temporary fix, but hopefully, it will suffice until a more permanent solution becomes available.

**Comment by rickiewars** [comment]
My previous temporary solution had some inconsistencies and bugs, which I've now fixed. I’d like to share the final version of this temporary workaround.

In this approach, I recursively add metadata to the child routes to avoid any inconsistencies. However, changes to the route.js files are still not automatically detected, so the Vite dev server will need to be restarted for those changes to take effect.

Add the following to your `vite.config.ts` file under `plugins.VueRouter`:
```ts
async beforeWriteFiles(rootRoute) {
  const config = {
    src: './resources/pages',
    routeOverrideFile: 'route.js',
    forcedRecursive: true
  }

  const iterateRoutes = (
    route: EditableTreeNode,
    callback: (route: EditableTreeNode) => Promise<void> | void
  ) => {
    callback(route)
    if (route.children) {
      route.children.forEach((childRoute: EditableTreeNode) => {
        iterateRoutes(childRoute, callback)
      })
    }
  }

  type ExtendRouteData = {
    route: EditableTreeNode
    data: any
  }
  const promises: Array<Promise<ExtendRouteData | null>> = []

  const loadRouteGroupOverrides = async (route: EditableTreeNode) => {
    const pagesDir = fileURLToPath(new URL(config.src, import.meta.url))
    const routeFile = path.join(pagesDir, route.name, config.routeOverrideFile)
    promises.push(
      import(routeFile)
        .then((module) => Object({ route: route, data: module.default }))
        .catch(() => null)
    )
  }

  iterateRoutes(rootRoute, async (route: EditableTreeNode) => {
    if (route.isPassThrough) {
      loadRouteGroupOverrides(route)
    }
  })

  await Promise.all(promises).then((results: Array<ExtendRouteData | null>) => {
    results.forEach((result) => {
      if (!result) {
        return
      }

      if (result.data.meta) {
        if (config.forcedRecursive) {
          // Simple fix for root ('/') route not passing their meta to children
          iterateRoutes(result.route, (route: EditableTreeNode) => {
            route.addToMeta(result.data.meta)
          })
        } else {
          result.route.addToMeta(result.data.meta)
        }
      }

      // Override other properties if needed
    })
  })
}
```

I hope this proves helpful until a more permanent solution is available.

**Comment by a81n9** [comment]
I want to know how Nuxt handles this

**Comment by A81N9** [comment]
Does vue-router v5 support this feature?
